### PR TITLE
`constrained-generators`: introduce a hook for naming variables

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

You can use either the `name "x" -> x` view pattern or the `[var| x |]` quasi quoter. This greatly simplifies debugging specs.

For example, when you do this:
```haskell
leqPair :: Specification BaseFn (Int, Int)
leqPair = constrained $ \[var| p |] ->
  match p $ \[var| x |] [var| y |] ->
    x <=. y
```
Instead of getting a plan from `printPlan leqPair` that talks about variables `v0`, `v1`, and `v3` it instead looks like this:
```
Simplified spec:
  constrained $ \ p_3 ->
    let x_1 = Fst (ToGeneric p_3) in
    let y_0 = Snd (ToGeneric p_3) in assert $ LessOrEqual x_1 y_0
SolverPlan
  Dependencies:
    y_0 <- []
    x_1 <- [y_0]
    p_3 <- [y_0, x_1]
  Linearization:
    y_0 <- 
    x_1 <- assert $ LessOrEqual x_1 y_0
    p_3 <-
      assert $ Equal (Fst (ToGeneric p_3)) x_1
      assert $ Equal (Snd (ToGeneric p_3)) y_0
```

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/constrained-generators/constrained-generators.cabal [hunk 1]
```diff
@@ -55,7 +56,8 @@ library
         mtl,
         prettyprinter,
         QuickCheck ^>=2.14,
-        random
+        random,
+        template-haskell
```
Comment (TimSheard) [inline libs/constrained-generators/constrained-generators.cabal:60]: Do we really want to depend on Template Haskell?

File: libs/constrained-generators/src/Constrained/Syntax.hs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    TH.AppE (TH.AppE (TH.VarE $ TH.mkName "name") (TH.LitE $ TH.StringL x)) (TH.VarE $ TH.mkName x)
+
+var :: TH.QuasiQuoter
+var =
+  TH.QuasiQuoter
+    { TH.quotePat = mkNamed . head . words
+    , TH.quoteExp = mkNamedExpr . head . words
+    , TH.quoteDec = const $ fail "var should only be used at binding sites and in expressions"
+    , TH.quoteType = const $ fail "var should only be used at binding sites and in expressions"
+    }
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained/Syntax.hs]: Maybe a comment about how this relates to quasiquoted  variable in a lambda expression.

spec = constrained $ \ [var| abc |] -> ...

that in    [xxx| abc |]  the system looks for a Quasiquoter with name "xxx", and uses it to parse the string "abc" into internal Haskell syntax.

**Review by TimSheard — APPROVED** [review]
Maybe add an example using the view syntax?
Other wise looks like it will be quote useful.

File: libs/constrained-generators/constrained-generators.cabal (hunk 1 above)
Comment (MaximilianAlgehed) [reply libs/constrained-generators/constrained-generators.cabal:60]: I honestly think this TH dependency is fine _because_ we are using a very limited amount of TH to do a specific job. The things we are using have been practically the same for at least a whole major version (IIRC) so I don't expect this to introduce any significant maintainance burden in the future.

File: libs/constrained-generators/src/Constrained/Syntax.hs (hunk 2 above)
Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Syntax.hs]: I've added a clearer explanation

**Pull request merged**